### DISCUSSION TITLE
How to wait until target file is created

### DISCUSSION THREAD

**Discussion by escwxyz** [chhoumann/quickadd#419] [Q&A] [answered]
Hello,

I am using another command to create a file from outer API. I made a macro and add enough waiting time between two.

But the second command either threw error due to lack of the target file or won't exec.

Thanks

**Comment by chhoumann** [comment, marked as answer]
Hi @escwxyz, are you using JS to create the file or a QuickAdd command? Or something else entirely?

For JS: remember to `await` asynchronous processes.

**Reply by escwxyz** [reply]
Hi,

thanks for your reply. I am using an Obsidian command which is from another plugin.

**Reply by chhoumann** [reply]
It's not possible at this point, but you just gave me an idea for how we might add it to the plugin.

Once #420 is in, you should be able to do this.

Reason why it's not possible presently:
The lever that QuickAdd pulls in Obsidian to execute commands from other plugins doesn't say _when_ the command is finished. So there's no way (with that approach) to wait for it to finish.
Since this is the 'official' way to execute commands, that is how QuickAdd currently does it.
But I do think it would be cool to have an approach which waits for the command to finish before going on to the next, which is what #420 is about. And luckily, there's a way to do it. Although, it's not supported by Obsidian, so might need some finesse to get right.

**Reply by escwxyz** [reply]
> It's not possible at this point, but you just gave me an idea for how we might add it to the plugin.
> 
> Once #420 is in, you should be able to do this.
> 
> Reason why it's not possible presently: The lever that QuickAdd pulls in Obsidian to execute commands from other plugins doesn't say _when_ the command is finished. So there's no way (with that approach) to wait for it to finish. Since this is the 'official' way to execute commands, that is how QuickAdd currently does it. But I do think it would be cool to have an approach which waits for the command to finish before going on to the next, which is what #420 is about. And luckily, there's a way to do it. Although, it's not supported by Obsidian, so might need some finesse to get right.

Thanks for your detailed reply. I am not familiar with Obsidian's API nor your plugins code. But what I have thought of is to predefine some event hooks before & after user's command or code, to expose kind of conditional APIs.

**Reply by chhoumann** [reply]
Yeah, it will be event-based for sure. I don't think I'll expose it as API functionality, but rather have it be accessible through the macro UI. IMO that's simpler for users.

**Comment by chhoumann** [comment]
Coming back to close the loop here, because my answer above promised something that never shipped: #420 was folded into #118, which was eventually closed as not planned. The underlying limitation still stands - Obsidian's command API gives no signal for when a command finishes, so QuickAdd can't offer a generic "wait for the command to complete" macro step.

But the question you actually asked - wait until the target file exists - has a clean answer: replace the fixed Wait with a small user script step that runs the other plugin's command and then polls for the file:

```js
module.exports = async (params) => {
    params.app.commands.executeCommandById("other-plugin:create-file-command");

    const path = "Folder/Target note.md";
    const timeout = Date.now() + 10_000;
    while (!params.app.vault.getAbstractFileByPath(path)) {
        if (Date.now() > timeout) throw new Error(`Timed out waiting for ${path}`);
        await new Promise((r) => setTimeout(r, 100));
    }
};
```

The macro only continues to the next step once the file exists (or aborts on the timeout instead of hanging forever). Docs on adding a script step: https://quickadd.obsidian.guide/docs/UserScripts
